src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -154,84 +172,109 @@ generateQ qq@(Query (c:_) _)
             -- >    (ret = ReturnOfLatest x't x'd)
             -- > => x't x'd -> Aggregate ret
             --
-            -- The definition for ReturnOfLatest is in Icicle.Source.Type.Constraints, but is something like
+            -- The definition for ReturnOfLatest is in Icicle.Source.Type.Constraints,
+            -- but is something like
+            --
             -- > ReturnOfLatest Aggregate d = d
             -- > ReturnOfLatest Element   d = Array d
             --
-            require a $ CReturnOfLatest retDat (fromMaybe TemporalityPure tmpq) datq
-            let t'  = canonT
-                    $ Temporality TemporalityAggregate
-                    $ Possibility PossibilityPossibly retDat
-            with q' sq t' $ \a' -> Latest a' i
+            let cons' = require a (CReturnOfLatest retDat (fromMaybe TemporalityPure tmpq) datq) cons
+            let t'    = canonT
+                      $ Temporality TemporalityAggregate
+                      $ Possibility PossibilityPossibly retDat
+
+            let q''   = with cons' q' t' $ \a' -> Latest a' i
+            return (q'', sq, CheckContext env cons')
 
     -- >   group (Element k'p k'd) ~> Aggregate v'p v'd
     -- > : Aggregate (PossibilityJoin k'p v'p) (Group k'd v'd)
     GroupBy _ x
-     -> do  (x',sx) <- generateX x
-            (q',sq,tval) <- rest
+     -> do  (x', sx, ctxKey)                      <- generateX x ctx
+            (q', sq, tval, CheckContext env cons) <- rest (ctxKey { ctxEnv = qenv })
```

Comment:
so I guess here you're passing ctxKey in order to use its constraints, so that rest returns the union of all constraints. however, you can just append the constraints together at this level (like we do with the substitutions)